File: gto-support-picasso/src/main/java/org/ccci/gto/android/common/picasso/view/SimplePicassoImageView.java
package org.ccci.gto.android.common.picasso.view;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Build;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.widget.ImageView;

import com.squareup.picasso.Transformation;

import java.io.File;
import java.util.List;

@SuppressLint("AppCompatCustomView")
public class SimplePicassoImageView extends ImageView implements PicassoImageView {
    private boolean mInit = false;
    @NonNull
    private final Helper mHelper;

    public SimplePicassoImageView(@NonNull final Context context) {
        this(context, null);
    }

    public SimplePicassoImageView(@NonNull final Context context, @Nullable final AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public SimplePicassoImageView(@NonNull final Context context, @Nullable final AttributeSet attrs,
                                  final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mHelper = new Helper(this, attrs, defStyleAttr, 0);
        mInit = true;
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public SimplePicassoImageView(@NonNull final Context context, @Nullable final AttributeSet attrs,
                                  final int defStyleAttr, final int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        mHelper = new Helper(this, attrs, defStyleAttr, defStyleRes);
        mInit = true;
    }

    @Override
    public final void setPicassoFile(@Nullable final File file) {
        mHelper.setPicassoFile(file);
    }

    @Override
    public final void setPicassoUri(@Nullable final Uri uri) {
        mHelper.setPicassoUri(uri);
    }

    @Override
    public final void setPlaceholder(@DrawableRes final int placeholder) {
        mHelper.setPlaceholder(placeholder);
    }

    @Override
    public void setPlaceholder(@Nullable final Drawable placeholder) {
        mHelper.setPlaceholder(placeholder);
    }

    @Override
    public void addTransform(@NonNull final Transformation transform) {
        mHelper.addTransform(transform);
    }

    @Override
    public void setTransforms(@Nullable List<? extends Transformation> transforms) {
        mHelper.setTransforms(transforms);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        mHelper.onSizeChanged(w, h, oldw, oldh);
    }

    @Override
    public void setScaleType(ScaleType scaleType) {
        super.setScaleType(scaleType);
        if (mInit) {
            mHelper.setScaleType(scaleType);
        }
    }
}

File: gto-support-picasso/src/main/java/org/ccci/gto/android/common/picasso/view/SimplePicassoImageView.kt
package org.ccci.gto.android.common.picasso.view

import android.content.Context
import android.graphics.drawable.Drawable
import android.net.Uri
import android.util.AttributeSet
import androidx.annotation.DrawableRes
import androidx.appcompat.widget.AppCompatImageView
import com.squareup.picasso.Transformation
import java.io.File

open class SimplePicassoImageView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : AppCompatImageView(context, attrs, defStyleAttr), PicassoImageView {
    protected open val helper by lazy { PicassoImageView.Helper(this, attrs, defStyleAttr) }
    private val initialized = true

    override fun setPicassoFile(file: File?) = helper.setPicassoFile(file)
    override fun setPicassoUri(uri: Uri?) = helper.setPicassoUri(uri)
    override fun setPlaceholder(@DrawableRes placeholder: Int) = helper.setPlaceholder(placeholder)
    override fun setPlaceholder(placeholder: Drawable?) = helper.setPlaceholder(placeholder)
    override fun addTransform(transform: Transformation) = helper.addTransform(transform)
    override fun setTransforms(transforms: List<Transformation>?) = helper.setTransforms(transforms)
    override fun toggleBatchUpdates(enable: Boolean) = helper.toggleBatchUpdates(enable)

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        helper.onSizeChanged(w, h, oldw, oldh)
    }

    override fun setScaleType(scaleType: ScaleType) {
        super.setScaleType(scaleType)

        // HACK: setScaleType() could be called by a parent constructor, which can cause this statement to crash. The
        //       current logic handles several possible crashes, but doesn't handle all the possibilities subclasses
        //       could introduce.
        @Suppress("UNNECESSARY_SAFE_CALL")
        if (initialized) helper?.onSetScaleType()
    }
}
